File: ios/App/Sources/Core/Domain/NavCaddy/SessionContext.swift
import Foundation

/// Represents the short-term session context for conversation continuity.
///
/// Spec R6: Maintains current round, hole, last shot, and conversation history.
struct SessionContext: Codable, Hashable {
    let sessionId: String
    let currentRound: Round?
    let currentHole: Int?
    let lastShot: Shot?
    let lastRecommendation: String?
    let conversationHistory: [ConversationTurn]

    init(
        sessionId: String = UUID().uuidString,
        currentRound: Round? = nil,
        currentHole: Int? = nil,
        lastShot: Shot? = nil,
        lastRecommendation: String? = nil,
        conversationHistory: [ConversationTurn] = []
    ) {
        self.sessionId = sessionId
        self.currentRound = currentRound
        self.currentHole = currentHole
        self.lastShot = lastShot
        self.lastRecommendation = lastRecommendation
        self.conversationHistory = conversationHistory
    }

    /// Returns a new context with the conversation turn added
    func addingTurn(_ turn: ConversationTurn) -> SessionContext {
        var history = conversationHistory
        history.append(turn)

        // Keep only last 10 turns as per spec
        if history.count > 10 {
            history = Array(history.suffix(10))
        }

        return SessionContext(
            sessionId: sessionId,
            currentRound: currentRound,
            currentHole: currentHole,
            lastShot: lastShot,
            lastRecommendation: lastRecommendation,
            conversationHistory: history
        )
    }

    /// Returns a new context with all fields cleared
    static var empty: SessionContext {
        SessionContext()
    }

    /// Whether a round is currently active.
    var isRoundActive: Bool {
        currentRound != nil
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/models/SessionContext.kt
package caddypro.domain.navcaddy.models

import com.google.gson.annotations.SerializedName

/**
 * Domain model representing the current session context.
 *
 * Maintains short-term state for conversational follow-ups and context-aware responses.
 *
 * Spec reference: navcaddy-engine.md R6
 *
 * @property currentRound Active round if in progress
 * @property currentHole Current hole number if in a round
 * @property lastShot Most recent shot recorded
 * @property lastRecommendation Last recommendation provided to user
 * @property conversationHistory Recent conversation turns (typically last 10)
 */
data class SessionContext(
    @SerializedName("current_round")
    val currentRound: Round? = null,

    @SerializedName("current_hole")
    val currentHole: Int? = null,

    @SerializedName("last_shot")
    val lastShot: Shot? = null,

    @SerializedName("last_recommendation")
    val lastRecommendation: String? = null,

    @SerializedName("conversation_history")
    val conversationHistory: List<ConversationTurn> = emptyList()
) {
    init {
        currentHole?.let {
            require(it in 1..18) { "Current hole must be between 1 and 18" }
        }
    }

    companion object {
        /** Maximum number of conversation turns to retain */
        const val MAX_HISTORY_SIZE = 10

        /** Empty session context for initialization */
        val empty = SessionContext()
    }

    /**
     * Returns a new SessionContext with the turn added and history limited to MAX_HISTORY_SIZE.
     *
     * @param turn The conversation turn to add
     * @return New SessionContext with updated history
     */
    fun addingTurn(turn: ConversationTurn): SessionContext {
        val newHistory = (conversationHistory + turn).takeLast(MAX_HISTORY_SIZE)
        return copy(conversationHistory = newHistory)
    }
}
